### PR TITLE
ref(discover): Dataset-specific config to specify fields and filters

### PR DESCRIPTION
**Pull request by shruthilayaj**
Introduce a dataset-specific configuration that generates
the available fields, functions and filters for the dataset
so that the logic doesn't live inside QueryBuilder. Furthermore,
remove the multiple levels of inheritance and move the 
related functions inside QueryBuilder since a lot of
complexities are abstracted away by the dataset config.
Note, not removing the QueryFilter, QueryFields and QueryBase
class yet because they are inherited or called elsewhere.
Will be removed as a follow up.

### CODE REVIEW

File: src/sentry/search/events/builder.py [hunk 1]
```diff
@@ -179,6 +475,541 @@ def get_snql_query(self) -> Query:
...
+            alias = get_function_alias_with_columns(raw_function, arguments)
+
+        return (function, combinator, arguments, alias)
+
+    def get_public_alias(self, function: CurriedFunction) -> str:
+        """Given a function resolved by QueryBuilder, get the public alias of that function
+
+        ie. any_user_display -> any(user_display)
+        """
+        return self.function_alias_map[function.alias].field  # type: ignore
```
Comment (shruthilayaj) [inline src/sentry/search/events/builder.py:1011]:
```
src/sentry/search/events/builder.py:1011: error: Returning Any from function declared to return "str"
```
But I don't understand whyyy

File: src/sentry/search/events/builder.py [hunk 2]
```diff
@@ -161,6 +230,229 @@ def validate_having_clause(self) -> None:
...
+        """
+        conditions = []
+
+        # start/end are required so that we can run a query in a reasonable amount of time
+        if "start" not in self.params or "end" not in self.params:
+            raise InvalidSearchQuery("Cannot query without a valid date range")
+
+        start: datetime
+        end: datetime
+        start, end = self.params["start"], self.params["end"]  # type: ignore
```
[the remark this answers is not in the document]
Comment (wmak) [reply src/sentry/search/events/builder.py:276]: There's a backlog task to change the loose typing on the params to a dataclass which will clean this up significantly

File: src/sentry/search/events/builder.py (hunk 1 above)
Comment (wmak) [reply src/sentry/search/events/builder.py:1011]: I think its cause `function.alias` might not exist, and `function_alias_map` might not have that key

**Review by wmak — APPROVED** [review]
LGTM, some nitpicks but overall a good direction for QueryBuilder

File: src/sentry/api/endpoints/organization_events_trends.py [hunk 3]
```diff
@@ -173,7 +173,7 @@ def resolve_trend_columns(
             avg_range_1 = aggregate_range_1
             avg_range_2 = aggregate_range_2
 
-        t_test = query.resolve_division(
+        t_test = query.config.resolve_division(
```
[the remark this answers is not in the document]
Comment (shruthilayaj) [reply src/sentry/api/endpoints/organization_events_trends.py]: 2c4b4e8bc11372ea616aa68d21746e2146d890b3

File: src/sentry/search/events/dataset.py [hunk 4]
```diff
@@ -0,0 +1,1475 @@
...
+from sentry.models import Environment, ProjectTeam, Release, SemverFilter
+from sentry.models.group import Group
+from sentry.models.project import Project
+from sentry.models.transaction_threshold import (
+    TRANSACTION_METRICS,
+    ProjectTransactionThreshold,
+    ProjectTransactionThresholdOverride,
+)
+from sentry.search.events.builder import QueryBuilder
+from sentry.search.events.constants import (
```
[the remark this answers is not in the document]
Comment (shruthilayaj) [reply src/sentry/search/events/dataset.py:26]: good point, will make a follow up 👍

File: tests/sentry/snuba/test_discover.py [hunk 5]
```diff
@@ -3215,7 +3215,7 @@ def test_project_in_condition_with_or(self):
         project2 = self.create_project(organization=self.organization)
         event_data = load_data("transaction", timestamp=before_now(seconds=3))
         self.store_event(data=event_data, project_id=project2.id)
-        expected = [self.project.slug, project2.slug]
+        expected = list(sorted([self.project.slug, project2.slug]))
```
[the remark this answers is not in the document]
Comment (shruthilayaj) [reply tests/sentry/snuba/test_discover.py]: removed!

File: src/sentry/search/events/builder.py [hunk 6]
```diff
@@ -9,14 +13,54 @@
...
+from sentry.search.events.filter import ParsedTerm, ParsedTerms, QueryFilter
 from sentry.search.events.types import HistogramParams, ParamsType, SelectType, WhereType
-from sentry.utils.snuba import Dataset
+from sentry.utils.dates import outside_retention_with_modified_start, to_timestamp
+from sentry.utils.snuba import Dataset, QueryOutsideRetentionError, resolve_column
+from sentry.utils.validators import INVALID_ID_DETAILS, INVALID_SPAN_ID, WILDCARD_NOT_ALLOWED
 
 
-class QueryBuilder(QueryFilter):  # type: ignore
+class QueryBuilder:
```
[the remark this answers is not in the document]
Comment (shruthilayaj) [reply src/sentry/search/events/builder.py:63]: 9d4a02bb746737fd5f7213018b3758cb58cb79ca

File: src/sentry/search/events/builder.py [hunk 7]
```diff
@@ -38,28 +82,69 @@ def __init__(
...
+        if self.dataset in [Dataset.Discover, Dataset.Transactions, Dataset.Events]:
+            self.config = DiscoverDatasetConfig(self)
+        else:
+            raise NotImplementedError(f"Data Set configuration not found for {self.dataset}.")
+
+        field_alias_converter = self.config.field_alias_converter
+
+        function_converter = self.config.function_converter
+        for alias, name in FUNCTION_ALIASES.items():
+            function_converter[alias] = function_converter[name].alias_as(alias)
```
Comment (Zylphrex) [inline src/sentry/search/events/builder.py:142]: Function aliases should probably live within the config as it's dataset specific.

**Review by Zylphrex — APPROVED** [review]

**Pull request merged**